extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java

```diff
@@ -95,8 +114,214 @@ public void handle(RoutingContext routingContext) {
+
+                final CloudEvent<?> inputCloudEvent;
+                final Object input;
+                if (invoker.hasInput()) {
+                    if (binaryCE) {
+                        inputCloudEvent = new HeaderCloudEventImpl<>(
+                                httpRequest.headers(),
+                                bodyBuff,
+                                inputCeDataType != null ? inputCeDataType : innerInputType,
+                                mapper,
+                                reader);
+                    } else {
+                        inputCloudEvent = new JsonCloudEventImpl<>(
+                                structuredPayload,
+                                inputCeDataType != null ? inputCeDataType : innerInputType,
+                                mapper,
+                                reader);
+                    }
+                    if (inputCeDataType == null) {
+                        // we need to unwrap user data from CloudEvent
+                        input = inputCloudEvent.data();
+                    } else {
+                        // user is explicitly handling CloudEvent
+                        input = inputCloudEvent;
+                    }
+                } else {
+                    input = inputCloudEvent = null;
+                }
+
+                final Consumer<Object> sendOutput = output -> {
+                    try {
+                        if (!invoker.hasOutput()) {
+                            routingContext.response().setStatusCode(204);
+                            routingContext.response().end();
+                            return;
+                        }
+
+                        final CloudEvent<?> outputCloudEvent;
+                        if (outputCeDataType == null) {
+                            // we need to wrap user data into CloudEvent
+                            String type = (String) invoker.getBindingContext().get(RESPONSE_TYPE);
+                            String source = (String) invoker.getBindingContext().get(RESPONSE_SOURCE);
+                            CloudEventBuilder builder = CloudEventBuilder.create()
+                                    .specVersion(CloudEvent.SpecVersion.V1)
+                                    .id(getResponseId())
+                                    .type(type)
+                                    .source(source)
+                                    .extensions(Collections.emptyMap());
+
+                            if (byte[].class.equals(innerOutputType)) {
+                                outputCloudEvent = builder.build((byte[]) output, "application/octet-stream");
+                            } else {
+                                outputCloudEvent = builder.build(output);
+                            }
+                        } else {
+                            // user is explicitly returning CloudEvent
+                            outputCloudEvent = (CloudEvent<?>) output;
```

Comment:
@mswiderski right here I should set defaults for mandatory CE attributes.

Reply:
perfect